### PR TITLE
OpenSSL ByteBuffer BIO

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
Currently Netty utilizes BIO_new_bio_pair so we can control all FD lifetime and event notification but delegates to OpenSSL for encryption/decryption. The current implementation sets up a pair of BIO buffers to read/write encrypted/plaintext data. This approach requires copying of data from Java ByteBuffers to native memory BIO buffers, and also requires both BIO buffers to be sufficiently large to hold application data. If direct ByteBuffers are used we can avoid coyping to/from the intermediate BIO buffer and just read/write directly from the direct ByteBuffer memory. We still need an internal buffer because OpenSSL may generate write data as a result of read calls (e.g. handshake, alerts, renegotiation, etc..), but this buffer doesn't have to be be large enough to hold application data.

Modifications:
- Take advantage of the new ByteBuffer based BIO provided by netty-tcnative instead of using BIO_read and BIO_write.

Result:
Less copying and lower memory footprint requirement per TLS connection.

### CODE REVIEW

**Comment by Scottmitch** [comment]
- only the last commit is part of this PR
- this PR depends upon https://github.com/netty/netty-tcnative/pull/216

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 1]
```diff
@@ -812,9 +763,18 @@ public final SSLEngineResult unwrap(
             }
 
             if (len < SslUtils.SSL_RECORD_HEADER_LENGTH) {
-                return new SSLEngineResult(BUFFER_UNDERFLOW, getHandshakeStatus(), 0, 0);
-            }
-
+                return newResult(BUFFER_UNDERFLOW, 0, 0, status);
+            }
+
+            // TODO(scott): limiting to a single packet can result in thrashing in the pipeline to decode application
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @normanmaurer - this is meant to be a discussion point.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 2]
```diff
@@ -845,9 +820,18 @@ public final SSLEngineResult unwrap(
             }
 
             if (len < SslUtils.SSL_RECORD_HEADER_LENGTH) {
-                return new SSLEngineResult(BUFFER_UNDERFLOW, getHandshakeStatus(), 0, 0);
-            }
-
+                return newResultMayFinishHandshake(BUFFER_UNDERFLOW, status, 0, 0);
+            }
+
+            // TODO(scott): limiting to a single packet can result in thrashing in the pipeline to decode application
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @normanmaurer - this is meant to be a discussion point.

File: pom.xml [hunk 3]
```diff
@@ -244,7 +244,7 @@
     <!-- Fedora-"like" systems. This is currently only used for the netty-tcnative dependency -->
     <os.detection.classifierWithLikes>fedora</os.detection.classifierWithLikes>
     <tcnative.artifactId>netty-tcnative</tcnative.artifactId>
-    <tcnative.version>2.0.0.Beta1</tcnative.version>
+    <tcnative.version>2.0.0.Final-SNAPSHOT</tcnative.version>
```
Comment (Scottmitch) [inline pom.xml]: this must change

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 4]
```diff
@@ -727,26 +665,63 @@ public final SSLEngineResult wrap(
...
+            } finally {
+                SSL.bioClearByteBuffer(networkBIO);
+                if (bioReadCopyBuf == null) {
+                    dst.position(dst.position() + bytesProduced);
+                } else {
+                    assert bioReadCopyBuf.readableBytes() <= dst.remaining() : "The destination buffer " + dst +
+                            " didn't have enough remaining space to hold the encrypted content in " + bioReadCopyBuf;
+                    dst.put(bioReadCopyBuf.internalNioBuffer(bioReadCopyBuf.readerIndex(),
+                                                      bioReadCopyBuf.readerIndex() + bytesProduced));
+                    bioReadCopyBuf.release();
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @Scottmitch can we ensure we also release when the `assert` fails. Just in case :)

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 5]
```diff
@@ -845,9 +820,18 @@ public final SSLEngineResult unwrap(
...
-
+                return newResultMayFinishHandshake(BUFFER_UNDERFLOW, status, 0, 0);
+            }
+
+            // TODO(scott): limiting to a single packet can result in thrashing in the pipeline to decode application
+            // data. The javadocs elude to attempting to decode one complete SSL/TLS network packet [1], but I don't
+            // think this is meant to enforce a strict maximum of 1. The SslHandler's decode method already loops
+            // through the data to only give "full packets" to the SSLEngine implementations, we duplicate that
+            // logic here, and OpenSSL will also know how to delimit data. Can we remove this restriction which causes
+            // unnecessary overhead?
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @Scottmitch I think we can remove this via a constructor option ? Like strict and non-strict mode ?

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 6]
```diff
@@ -1387,14 +1349,27 @@ public final synchronized void beginHandshake() throws SSLException {
...
+                int status;
+                if ((status = SSL.renegotiate(ssl)) != 1 || (status = SSL.doHandshake(ssl)) != 1) {
+                    int err = SSL.getError(ssl, status);
+                    switch (err) {
+                        case SSL.SSL_ERROR_WANT_READ:
+                        case SSL.SSL_ERROR_WANT_WRITE:
+                            // If the internal SSL buffer is small it is possible that doHandshake may "fail" because
+                            // there is not enough room to write, so we should try to wrap.
+
+                            // TODO(scott): there may be a OpenSSL bug here. If the internal SSL buffer is < 33 bytes
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @Scottmitch can you open an issue and link it here ?

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java [hunk 7]
```diff
@@ -790,6 +796,7 @@ private void wrapNonAppData(ChannelHandlerContext ctx, boolean inUnwrap) throws
                 out.release();
             }
         }
+        return false;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: @Scottmitch like discussed before I think we should factor this out in a separate PR.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 4 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: I don't think this is necessary. The assert signifies a programatic error and was added to aid development and unit tests. Do you have a suggestion which doesn't introduce runtime checks or additional try/catch?

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 5 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: @normanmaurer - great! I noticed this was thrashing a lot during my testing. let me break this out into another PR.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 6 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: https://github.com/openssl/openssl/issues/2551

File: handler/src/main/java/io/netty/handler/ssl/SslHandler.java (hunk 7 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/SslHandler.java]: https://github.com/netty/netty/pull/6311

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 6 above)
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: false alarm ... issue was pre-existing but has been fixed in this PR.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java [hunk 8]
```diff
@@ -612,108 +509,162 @@ public final SSLEngineResult wrap(
...
+                        if (handshakeException == null) {
+                            status = handshake();
+
+                            if (renegotiationPending && status == FINISHED) {
+                                // If renegotiationPending is true that means when we attempted to start renegotiation
+                                // the BIO buffer didn't have enough space to hold the HelloRequest which prompts the
+                                // client to initiate a renegotiation. At this point the HelloRequest has been written
+                                // so we can actually start the handshake process.
+                                renegotiationPending = false;
+                                SSL.setState(ssl, SSL.SSL_ST_ACCEPT);
```
Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: The OpenSSL team has advised against doing this. I have fixed the current approach in this PR and will investigate removal in of `SSL.SSL_ST_ACCEPT` in https://github.com/netty/netty/issues/6320.

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java (hunk 4 above)
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java]: yeah... its ok

File: handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java [hunk 9]
```diff
@@ -85,6 +85,17 @@ public Boolean run() {
                     return SystemPropertyUtil.getBoolean("jdk.tls.rejectClientInitiatedRenegotiation", false);
                 }
             });
+
+    private static final int DEFAULT_BIO_NON_APPLICATION_BUFFER_SIZE =
+            AccessController.doPrivileged(new PrivilegedAction<Integer>() {
+                @Override
+                public Integer run() {
+                    return Math.max(1,
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java]: why 1 and not 0 (which will use the default then) ?

Comment (Scottmitch) [inline handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslContext.java]: I removed support for the default at the tcnative level, and disallowed setting 0. Users should explicitly set the size they want. Default of 16k is likely way to big anyways.

File: pom.xml (hunk 3 above)
Comment (Scottmitch) [inline pom.xml]: @normanmaurer will update this before the next release

**Comment by Scottmitch** [comment]
4.1 (d06990f) 4.0 (00cd6d2)

**Comment by normanmaurer** [comment]
LGTM... Thanks!
